### ISSUE TITLE
Abstract doesn't find underlying operator overload

### ISSUE THREAD

**Issue by ibilon** [HaxeFoundation/haxe#4656]
Hi,

I'm trying to do operator overloading using the underlying type operation.

I've got an extern class (hxcpp) like this
```
@:native("Example") extern class Underlying {
  @:native("operator+") public function operator_plus (value:Int) : Underlying;
}
```
and an abstract over it
```
abstract Example (Underlying) from Underlying {
  @:op(A + B) public function operator_plus (value:Int) : Underlying;
}
```
use case
```
var e : Example (get it somewhere)
e = e + 10;
```
gives `Cannot add Underlying and Int`.

Am I using this wrong? Or is this possible only for special underlying types?

Also quick question, can this http://haxe.org/manual/types-numeric-operators.html be used as a list of available operator for abstract operator overloading?

**Comment by Simn** [comment]
When you have an `@:op` field without expression the compiler checks if the operation is allowed on the underlying type, which is not the case here. This is a restriction we originally put into place when implementing operator overloading on abstracts.

What you're doing looks like you'd be better of using a `@:coreType` abstract to begin with:

```haxe
@:coreType
abstract Example {
	@:op(A + B) public function operator_plus (value:Int) : Example;
}

class Main {
	static function main() {
		var e:Example = null;
		e = e + 10;
	}
}
```

Is there any reason you have to define that abstract over an extern class?

**Comment by ibilon** [comment]
I see, tried adding `@:op(A + B)` to the extern's function, but I guess that's not how the check is done :)

I want to use a c++ class which has some operator overloading,
so my first idea was to make an extern class and an abstract over it,
but I'm open to any other solution.

**Comment by Simn** [comment]
Try using `@:coreType` abstracts like I suggested.

**Comment by nadako** [comment]
Other option is to use `untyped __cpp__` magic:
```haxe
extern class Example {
    inline function add(other:Example):Example {
        return untyped __cpp__("{0} + {1}", this, other);
    }
}
```

**Comment by ibilon** [comment]
The `@:coreType` requires that all my other functions to have a body, and I don't think it uses `@:native` the same way than an extern class (for `@:native("new Example") for instance),
and I can't do @nadako's solution because a lot of the class' functions are overloaded.

Also looking at the errors `@:include` is probably ignored.

**Comment by Simn** [comment]
I don't really know what to do with this. We don't support operator overloading on extern classes and I don't think we should remove the operator checks regarding the underlying type.

I really think `@:coreType` is the solution here. This works as expected:

```haxe
@:coreType
abstract Example {
	@:op(A + B) public function operator_plus (value:Int) : Example;
}

@:keep
class Main {
	static function main() {
		var e:Example = null;
		var e = e + 1;
	}
}
```

**Comment by ibilon** [comment]
For the operators, but it break all other extern functions :/

**Comment by Simn** [comment]
How so?

**Comment by ibilon** [comment]
It needs a function body, but because of overloading I can't manually use `__cpp__`.

I'd want to do something like this:
```haxe
@:coreType
abstract Example
{
    @:extern @:native("new Create") public function create () : Example;
    
    @:overload(function (i:Int) : Void {})
    @:extern @:native("Show") public function show () : Void;
    
    @:op(A+B) public function operator_plus (i:Int) : Example;
}
````

**Comment by ibilon** [comment]
I just realized that operators can't be overloaded, wouldn't make sense,
so I can make an inline function to forward to the underlying type.

I'm still blocked by the overloaded static functions,
but that's another issue https://github.com/HaxeFoundation/haxe/issues/3254
